### ISSUE TITLE
When using `dependsOn`, inputs seems to get overridden and target does not get cache hit

### ISSUE THREAD

**Issue by KyleBastien** [nrwl/nx#11700]
## Current Behavior

Currently if you have a target that has been configured to use `dependsOn` and run a cacheable operation, then the cacheable operation's `inputs` is not respected, and instead the `inputs` of the original task are used. This causes a cache miss for the task, even though it should be cached.

For example, let's say you have a build target which is configured like the following, with a `dependsOn`:

```json
"build": {
  "executor": "@nrwl/node:webpack",
  "outputs": ["{options.outputPath}"],
  "dependsOn": ["compileProto"],
  "options": {
    "outputPath": "dist/packages/test-app",
    "main": "packages/test-app/src/main.ts",
    "tsConfig": "packages/test-app/tsconfig.app.json",
    "assets": ["packages/test-app/src/assets"]
  },
  "configurations": {
    "production": {
      "optimization": true,
      "extractLicenses": true,
      "inspect": false,
      "fileReplacements": [
        {
          "replace": "packages/test-app/src/environments/environment.ts",
          "with": "packages/test-app/src/environments/environment.prod.ts"
        }
      ]
    }
  }
},
```

And your `compileProto` target is configured like this:

```json
"compileProto": {
  "executor": "@nrwl/workspace:run-commands",
  "inputs": ["{projectRoot}/src/app/test-folder-proto/**/*.proto"],
  "outputs": ["{projectRoot}/src/app/test-folder-output"],
  "options": {
    "commands": [
      {
        "command": "timeout 5"
      },
      {
        "command": "echo \"yay\""
      }
    ],
    "parallel": false
  }
}
```

We can see that `compileProto` is confused to only look at specific `inputs` (the folder where our proto files are stored in this package). And it outputs to a specific folder of our package as well.

Now if you run `nx run test-app:compileProto`, this behaves correctly, it will cache the results unless the files in `{projectRoot}/src/app/test-folder-proto/**/*.proto` have changed.

But if you run `nx run test-app:build` this does not behave correctly, it will always re-run the `compileProto` target as long as any file in `test-app` has changed. Instead of looking at the `compileProto` inputs, I believe `inputs` is getting overridden here to `build`'s inputs.

Note that this behavior is different than if you use `run-commands` to configure this target. If you configure `build` like this:

```json
"build-other": {
  "executor": "@nrwl/workspace:run-commands",
  "options": {
    "commands": [
      {
        "command": "nx run test-app:compileProto"
      },
      {
        "command": "nx run test-app:build"
      }
    ],
    "parallel": false
  }
},
```

Then this works correctly. It will only re-run compileProto if the `inputs` specified in `compileProto` have changed.

## Expected Behavior

I'd expect that using `dependsOn` has the same behavior as `run-commands`, that is that the `compileProto` target is only re-executed when the inputs of that target change.

## Steps to Reproduce

https://github.com/KyleBastien/nx-inputs-example

You can see a repro of this issue in the above repo.

You can test this by doing the following:

```bash
git clone https://github.com/KyleBastien/nx-inputs-example
npm install
# Cache the compileProto build
npx nx run test-app:compileProto
# Make a change to main.ts in test-app, then re-run the following
npx nx run test-app:compileProto
# Notice that the above outputs the cached build that we did earlier
# Now, run build to cache the test-app build
npx nx run test-app:build
# Make another change to main.ts in test-app, then re-run the following
npx nx run test-app:build
# Notice how this re-ran compileProto, even though compileProto's inputs didn't change
# Now, run build-other to cache the test-app build-other output
npx nx run test-app:build-other
# Make another change to main.ts in test-app, then re-run the following
npx nx run test-app:build-other
# Notice how this did not re-run compileProto, because compileProto's inputs didn't change
```

### Failure Logs
N/A

### Environment

```
>  NX   Report complete - copy this into the issue template

   Node : 16.15.1
   OS   : win32 x64
   npm  : 8.11.0

   nx : 14.5.10
   @nrwl/angular : Not Found
   @nrwl/cypress : Not Found
   @nrwl/detox : Not Found
   @nrwl/devkit : 14.5.10
   @nrwl/eslint-plugin-nx : 14.5.10
   @nrwl/express : Not Found
   @nrwl/jest : 14.5.10
   @nrwl/js : 14.5.10
   @nrwl/linter : 14.5.10
   @nrwl/nest : Not Found
   @nrwl/next : Not Found
   @nrwl/node : 14.5.10
   @nrwl/nx-cloud : Not Found
   @nrwl/nx-plugin : Not Found
   @nrwl/react : Not Found
   @nrwl/react-native : Not Found
   @nrwl/schematics : Not Found
   @nrwl/storybook : Not Found
   @nrwl/web : Not Found
   @nrwl/workspace : 14.5.10
   typescript : 4.7.4
   ---------------------------------------
   Local workspace plugins:
   ---------------------------------------
   Community plugins:
```

**Comment by meshde** [comment]
The problem seems to be in this code:
https://github.com/nrwl/nx/blob/cd3352aaa2f99eabaf31ebaabc09ef981f4241b9/packages/nx/src/hasher/hasher.ts#L452-L453

This is the code that calculates the hash of the input files of a task. In this case when calling `nx run test-app:build`, there are two tasks; the root task itself (`build`) and the task it depends on (`compileProto`). 

The problem here is that the `hashProjectFileset` function returns the same hash for both the tasks even though the tasks have different input file globs. This is because the `mapKey` turns out to be the same for both tasks.

So what really happens is:
1. `hashProjectFileset` gets called for the root task (`build`).
2. The gets called with the input file set pattern of `{projectRoot}/**/*` (the default input pattern).
3. The hash gets correctly calculated considering all the files in the project.
4. The hash gets stored in `this.filesetHashes` under the key `"test-app:$filesets:default"` (map key).
5. `hashProjectFileset` gets called for the depends on task (`compileProto`).
6. The map key in this case also turns out to be `"test-app:$filesets:default"`  even though input file set pattern is different (`{projectRoot}/src/app/test-folder-proto/**/*.proto`).
7. The hash for this particular input glob is not calculated but instead the hash already stored in `this.filesetHashes["test-app:$filesets:default"]` gets returned.
8. This is why `compileProto` gets run again even though the change is in a file outside it's input file pattern.


I believe the following change should fix the issue, considering how the value of `filesetPatterns` is correctly passed to this function:

```diff
-    const mapKey = `${projectName}:$filesets:${namedInput}`;
+    const mapKey = `${projectName}:$filesets:${namedInput}:${filesetPatterns}`;
```


IMO adding the `filesetPatterns` to the key makes sense as the hash would be different for different file set patterns.

Happy to raise a PR to fix this.

**Comment by skrtheboss** [comment]
Should be solved by https://github.com/nrwl/nx/pull/12162

**Comment by FrozenPandaz** [comment]
Fixed by https://github.com/nrwl/nx/pull/12162 but https://github.com/nrwl/nx/pull/11860 was also a valid fix. The fix has been released in `14.7.11`

**Comment by KyleBastien** [comment]
Can verify https://github.com/KyleBastien/nx-inputs-example now shows the correct behavior with 14.7.11.

Thanks @FrozenPandaz, @meshde, and @vsavkin for the fix.
